client/my-sites/plans-features-main/hooks/use-max-plan-upgrade-credits.ts

```diff
@@ -15,30 +13,37 @@ interface Props {
  * @returns {number} The maximum amount of credits possible for a given set of plans
  */
 export function useMaxPlanUpgradeCredits( { siteId, plans }: Props ): number {
-	const plansDetails = useSelector( ( state ) =>
-		plans.map( ( planName ) => ( {
-			isPlanAvailableForPurchase: isPlanAvailableForPurchase( state, siteId ?? 0, planName ),
-			planDiscountedRawPrice: getPlanDiscountedRawPrice( state, siteId ?? 0, planName ),
-			sitePlanRawPrice: getSitePlanRawPrice( state, siteId ?? 0, planName ),
-		} ) )
-	);
+	const planAvailabilityForPurchase = useCheckPlanAvailabilityForPurchase( {
+		planSlugs: plans,
+		siteId,
+	} );
+	const pricing = Plans.usePricingMetaForGridPlans( {
+		siteId,
+		planSlugs: plans,
+		storageAddOns: null,
+		coupon: undefined,
+		useCheckPlanAvailabilityForPurchase,
+		withProratedDiscounts: true,
+	} );
 
-	if ( ! siteId ) {
+	if ( ! siteId || ! pricing ) {
 		return 0;
 	}
 
-	const creditsPerPlan = plansDetails.map(
-		( { isPlanAvailableForPurchase, planDiscountedRawPrice, sitePlanRawPrice } ) => {
-			if ( ! isPlanAvailableForPurchase ) {
-				return 0;
-			}
-			if ( typeof planDiscountedRawPrice !== 'number' || typeof sitePlanRawPrice !== 'number' ) {
-				return 0;
-			}
+	const creditsPerPlan = plans.map( ( planSlug ) => {
+		const discountedPrice = pricing?.[ planSlug ]?.discountedPrice.full;
```

Comment:
`Non-blocking / Nitpick`:

It was unusual to me that we needed to do optional chaining on the pricing object here, given that:

1. we know the pricing object exists after evaluating the condition on line 29.
2. we also know that all planSlugs fed into `usePricingMetaForGridPlans` have a 1:1 mapping in the hook's return value

I would've expected typescript to be a bit smarter about the situation. I then saw that the inferred type for pricing was

```
const pricing: {
    [planSlug: string]: Plans.PricingMetaForGridPlan;
}
```

I went down a bit of a rabbit hole trying to update the `planSlug` key from a `string` type to a `PlanSlug` type, and it looks like it opens up a can of worms 😆

I'll leave things as is for now because it doesn't seem like it's worth fixing, but I wanted to note this if anyone has similar questions when looking at this PR.

Reply:
created this issue https://github.com/Automattic/wp-calypso/issues/94475